java/src/main/java/cucumber/runtime/java/ObjectFactoryLoader.java

```diff
@@ -53,6 +55,46 @@ public static ObjectFactory loadObjectFactory(ClassFinder classFinder, String ob
         return objectFactory;
     }
 
+    private static io.cucumber.core.backend.ObjectFactory loadSingleObjectFactory(final Reflections reflections, final ServiceLoader<io.cucumber.core.backend.ObjectFactory> loader) {
+        final Map<String, io.cucumber.core.backend.ObjectFactory> objectFactories = loadObjectFactories(reflections, loader);
+        final io.cucumber.core.backend.ObjectFactory result;
+        if (objectFactories.isEmpty()) {
+            throw new NoInstancesException(io.cucumber.core.backend.ObjectFactory.class);
+        } else if (objectFactories.size() > 1) {
+            throw new TooManyInstancesException(objectFactories.values());
+        } else {
+            result = objectFactories.values().iterator().next();
+        }
+        return result;
+    }
+
+    private static Map<String, io.cucumber.core.backend.ObjectFactory> loadObjectFactories(final Reflections reflections, final ServiceLoader<io.cucumber.core.backend.ObjectFactory> loader) {
+        final Map<String, io.cucumber.core.backend.ObjectFactory> knowObjectFactories = new HashMap<>();
+        final List<URI> packages = asList(URI.create("classpath:cucumber/runtime"));
+        loadObjectFactories(reflections.instantiateSubclasses(io.cucumber.core.backend.ObjectFactory.class, packages, new Class[0], new Object[0]), knowObjectFactories);
+        loadObjectFactories(loader, knowObjectFactories);
+        return knowObjectFactories;
+    }
+
+    private static io.cucumber.core.backend.ObjectFactory loadSelectedObjectFactory(final Reflections reflections, final ServiceLoader<io.cucumber.core.backend.ObjectFactory> loader, final ClassFinder classFinder, final String objectFactoryClassName) throws ClassNotFoundException {
+        final Map<String, io.cucumber.core.backend.ObjectFactory> objectFactoriesByServiceLoader = new HashMap<>();
+        loadObjectFactories(loader, objectFactoriesByServiceLoader);
+        final io.cucumber.core.backend.ObjectFactory result;
+        if (objectFactoriesByServiceLoader.containsKey(objectFactoryClassName)) {
```

Comment:
If you know the class name you can use https://docs.oracle.com/javase/7/docs/api/java/util/ServiceLoader.html#load(java.lang.Class) instead. This avoids instantiation of all the services.

Reply:
sry? but I do not know the class? We know the interface?
This is only a workaround to check if service loader and old reflection found the same Classes or found different. Can happen if you migrate a ObjectFactory to ServiceLoader and than put a old one into class path. Okay I can change the impl. so we use ServiceLoader only if Reflection not found one? Do you think this is the better way?

Reply:
We do know the class name. It's `objectFactoryClassName`. You can use that to get the class.

We should give preference to the new system over the old one. So if the `ServiceLoader` does not give us any results we fall back to reflection.

Reply:
Reason being that we deprecated the old solution. So it should not be used when not needed.